src/events.ts

```diff
@@ -0,0 +1,86 @@
+import { FrameElement } from "./elements/frame_element"
+import { StreamElement } from "./elements/stream_element"
+
+import { FrameViewRenderOptions } from "./core/frames/frame_view"
+import { PageViewRenderOptions } from "./core/drive/page_view"
+import { FormSubmission, FormSubmissionResult } from "./core/drive/form_submission"
+
+import { FetchResponse } from "./http/fetch_response"
+import { FetchRequest } from "./http/fetch_request"
+
+import { Action, TimingData, VisitFallback, Render } from "./core/types"
+
+export type TurboBeforeCacheEvent = CustomEvent
+export type TurboBeforeRenderEvent = CustomEvent<{ newBody: HTMLBodyElement } & PageViewRenderOptions>
+export type TurboBeforeVisitEvent = CustomEvent<{ url: string }>
+export type TurboClickEvent = CustomEvent<{ url: string; originalEvent: MouseEvent }>
+export type TurboFrameLoadEvent = CustomEvent
+export type TurboBeforeFrameRenderEvent = CustomEvent<{ newFrame: FrameElement } & FrameViewRenderOptions>
+export type TurboFrameRenderEvent = CustomEvent<{ fetchResponse: FetchResponse }>
+export type TurboLoadEvent = CustomEvent<{ url: string; timing: TimingData }>
+export type TurboRenderEvent = CustomEvent
+export type TurboReloadEvent = CustomEvent
+export type TurboVisitEvent = CustomEvent<{ url: string; action: Action }>
+
+export type TurboBeforeStreamRenderEvent = CustomEvent<{ newStream: StreamElement; render: Render }>
+
+export type TurboSubmitStartEvent = CustomEvent<{ formSubmission: FormSubmission }>
+export type TurboSubmitEndEvent = CustomEvent<
+  { formSubmission: FormSubmission } & { [K in keyof FormSubmissionResult]?: FormSubmissionResult[K] }
+>
+
+export type TurboFrameMissingEvent = CustomEvent<{ response: Response; visit: VisitFallback }>
+
+export type TurboBeforeFetchRequestEvent = CustomEvent<{
+  fetchOptions: RequestInit
+  url: URL
+  resume: (value: any) => void
+}>
+
+export type TurboBeforeFetchResponseEvent = CustomEvent<{
+  fetchResponse: FetchResponse
+}>
+
+export type TurboFetchRequestErrorEvent = CustomEvent<{
+  request: FetchRequest
+  error: Error
+}>
+
+interface TurboGlobalEventMap {
+  "turbo:before-cache": TurboBeforeCacheEvent
+  "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
```

Comment:
Is it possible to cut down on the duplication between lists by declaring `TurboGlobalEventMap` and `Turbo{HTML,}ElementEventMap` with `type` declarations, then making `TurboGlobalEventMap` union with `Turbo{HTML,}ElementEventMap`?

```diff
diff --git a/src/events.ts b/src/events.ts
index 0fd77d1..7a8ae2e 100644
--- a/src/events.ts
+++ b/src/events.ts
@@ -46,28 +46,19 @@ export type TurboFetchRequestErrorEvent = CustomEvent<{
   error: Error
 }>
 
-interface TurboGlobalEventMap {
+type TurboGlobalEventMap = {
   "turbo:before-cache": TurboBeforeCacheEvent
-  "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
-  "turbo:before-fetch-response": TurboBeforeFetchResponseEvent
-  "turbo:before-frame-render": TurboBeforeFrameRenderEvent
   "turbo:before-stream-render": TurboBeforeStreamRenderEvent
   "turbo:before-render": TurboBeforeRenderEvent
   "turbo:before-visit": TurboBeforeVisitEvent
   "turbo:click": TurboClickEvent
-  "turbo:fetch-request-error": TurboFetchRequestErrorEvent
-  "turbo:frame-load": TurboFrameLoadEvent
-  "turbo:frame-render": TurboFrameRenderEvent
-  "turbo:frame-missing": TurboFrameMissingEvent
   "turbo:load": TurboLoadEvent
   "turbo:render": TurboRenderEvent
   "turbo:reload": TurboReloadEvent
-  "turbo:submit-start": TurboSubmitStartEvent
-  "turbo:submit-end": TurboSubmitEndEvent
   "turbo:visit": TurboVisitEvent
-}
+} & TurboHTMLElementEventMap
 
-interface TurboElementEventMap {
+type TurboHTMLElementEventMap = {
   "turbo:before-frame-render": TurboBeforeFrameRenderEvent
   "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
   "turbo:before-fetch-response": TurboBeforeFetchResponseEvent
@@ -82,5 +73,5 @@ interface TurboElementEventMap {
 declare global {
   interface DocumentEventMap extends TurboGlobalEventMap {}
   interface WindowEventMap extends TurboGlobalEventMap {}
-  interface ElementEventMap extends TurboElementEventMap {}
+  interface HTMLElementEventMap extends TurboHTMLElementEventMap {}
 }
```